### PR TITLE
[SR-7338] Handle cookies in URLSession

### PR DESCRIPTION
**Pull request by mamabusi**
This PR implements storage of HTTP cookies received as part of the response from the server and setting of appropriate cookies on the URLRequest of the URLSession.

### CODE REVIEW

**Comment by pushkarnk** [comment]
@swift-ci please test

**Comment by alblue** [comment]
@swift-ci please test

**Comment by spevans** [comment]
@swift-ci please test

**Comment by mamabusi** [comment]
@ianpartridge Could you please review this PR.

File: Foundation/URLSession/URLSession.swift [hunk 1]
```diff
@@ -383,8 +383,8 @@ fileprivate extension URLSession {
         case url(URL)
     }
     func createConfiguredRequest(from request: URLSession._Request) -> URLRequest {
-        let r = request.createMutableURLRequest()
-        _configuration.configure(request: r)
+        var r = request.createMutableURLRequest()
+        r = _configuration.configure(request: r)
```
Comment (ianpartridge) [inline Foundation/URLSession/URLSession.swift]: Why not just `return _configuration.configure(request: r)`?

File: Foundation/URLSession/Configuration.swift [hunk 2]
```diff
@@ -100,19 +100,30 @@ internal extension URLSession._Configuration {
...
 // Configure NSURLRequests
 internal extension URLSession._Configuration {
-    func configure(request: URLRequest) {
+    func configure(request: URLRequest) -> URLRequest {
         var request = request
         httpAdditionalHeaders?.forEach {
             guard request.value(forHTTPHeaderField: $0.0) == nil else { return }
             request.setValue($0.1, forHTTPHeaderField: $0.0)
         }
+        request = setCookies(on: request)
```
Comment (ianpartridge) [inline Foundation/URLSession/Configuration.swift]: `return setCookies(on: request)`?

File: Foundation/URLSession/Configuration.swift [hunk 3]
```diff
@@ -100,19 +100,30 @@ internal extension URLSession._Configuration {
...
+        return request
     }
-    func setCookies(on request: URLRequest) {
+
+     func setCookies(on request: URLRequest) -> URLRequest {
+        var request = request
         if httpShouldSetCookies {
-            //TODO: Ask the cookie storage what cookie to set.
+            if let cookieStorage = self.httpCookieStorage, let cookies = cookieStorage.cookies(for: request.url!) {
```
Comment (ianpartridge) [inline Foundation/URLSession/Configuration.swift]: Please remove the force unwrap, e.g.

```swift
if let cookieStorage = self.httpCookieStorage,
   let url = request.url,
   let cookies = cookieStorage.cookies(for: url) {
    ...
}
```

**Comment by mamabusi** [comment]
@ianpartridge  Thanks for the suggestions. I have made the necessary changes an updated the branch.

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by maksimorlovich** [comment]
@spevans @mamabusi @ianpartridge Would it be possible to get this merged into 4.2 branch? I think this is an important piece of functionality that's missing from Swift 4.2 release.

Also, looks like this was partially already merged in via this PR:
https://github.com/apple/swift-corelibs-foundation/pull/1615

I think it was by accident? The new test cases made it to Swift 4.2 branch, but not the actual code...

**Comment by mamabusi** [comment]
@maksimorlovich Sure, I'll be creating a PR soon for the Swift 4.2 branch.

**Comment by maksimorlovich** [comment]
@mamabusi , I created a PR to Swift 4.2 branch here: https://github.com/apple/swift-corelibs-foundation/pull/1676 Please take a look!

**Pull request merged**